mls-rs/src/group/generation_authentication.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// Copyright by contributors to this project.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+use super::{ciphertext_processor::GroupStateProvider, framing::ContentType, secret_tree::KeyType};
+use crate::{client::MlsError, tree_kem::node::NodeIndex};
+use mls_rs_core::crypto::CipherSuiteProvider;
+
+/// Key generation authentication used when encrypting a group message.
+/// Defends against in-group forgery attacks where group members can replay messages sent
+/// by other members, see https://eprint.iacr.org/2025/554.
+#[cfg_attr(all(feature = "ffi", not(test)), safer_ffi_gen::ffi_type)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+#[repr(u8)]
+pub enum GenerationAuthMode {
+    /// No generation authentication.
+    #[default]
+    None,
+    /// Prepends the 4-byte generation in big endian as plaintext to PrivateMessage's
+    /// authenticated data.
+    ///
+    /// # Warning
+    ///
+    /// This sends the generation in the clear, allowing non-group members observing
+    /// payloads to order messages and learn how many messages a member has sent.
+    Plaintext,
+}
+
+impl GenerationAuthMode {
+    /// Prepend the key generation to `authenticated_data` based on the mode.
+    pub(super) fn prepend_generation<GS: GroupStateProvider, CS: CipherSuiteProvider>(
+        &self,
+        group_state: &mut GS,
+        cipher_suite_provider: &CS,
+        content_type: ContentType,
+        authenticated_data: &mut Vec<u8>,
```

Comment:
I'm on my phone, but I don't immediately see where this needs to be mutable?

As a rule of thumb, it's unusual to take a mutable reference as input. So it calls attention to itself 🙂

Reply:
Changed in newest commit that I hadn't pushed earlier.